Documentation Updates (#19)
nautobot/nautobot-app-device-onboarding: README.md

> # }

Reviewer: Flow is a bit awkward here, as the above example talks about putting things in `PLUGINS_CONFIG`, but then we go through the NAPALM settings and the nautobot-server migration below before returning to the available `Plugin Configuration Options`. Consider moving that section up to here?

Author: Good catch! - updated

### Preparation

To properly onboard a device, user needs to provide, at a minimum:
1. The Device's Site 
2. The Device's primary IP address or DNS Name

> For DNS Name Resolution to work, the instance of Nautobot must be able to resolve the name of the
> device to IP address.

If other attributes (`Platform`, `Device Type`, `Device Role`) are provided in the onboarding task, the plugin will use provided value for the onboarded device.
If `Platform`, `Device Type` and/or `Device Role` are not provided, the plugin will try to identify these information automatically and, based on the settings, it can create them in Nautobot as needed.
> If the Platform is provided, it must point to an existing Nautobot Platform. NAPALM driver of this platform will be used only if it is defined for the platform in Nautobot.
> To use a preferred NAPALM driver, either define it in Nautobot per platform or in the plugins settings under `platform_map`

#### SSH Autodetect

Plugin recognizes platform types with a Netmiko SSH Autodetect mechanism. The user will need to specify additional information for platforms where Netmiko's `ssh_autodetect` feature does not work. 

> [Here is the list](https://github.com/ktbyers/netmiko/blob/2dc032b64c3049d3048966441ee30a0139bebc81/netmiko/ssh_autodetect.py#L50) of platforms supported by `ssh_autodetect`.

Reviewer: Can we link to a specific version instead of a commit hash?
```suggestion
[Here is the list](https://github.com/ktbyers/netmiko/blob/v3.4.0/netmiko/ssh_autodetect.py#L50) of platforms supported by `ssh_autodetect`.
```

> The `nautobot-device-onboarding` plugin is compatible for devices that are supported by NAPALM. It is possible to create user's custom NAPALM driver and use them with a plugin.

Reviewer: This paragraph reads a bit awkwardly to me - maybe instead:

```suggestion
The `nautobot-device-onboarding` plugin can be used with any devices that are supported by NAPALM. Even custom NAPALM driver plugins can be used with a bit of effort.
```

Devices that are supported by NAPALM but are not running SSH or don't have support for `ssh_autodetect` will still work with this plugin, but will require some additional information in the onboarding task.

The table below shows which common platforms will be SSH auto-detected by default.

|Platform     |Platform Autodetect|
--------------|--------------------
Juniper/Junos | Yes (when running Netconf over SSH)|
Cisco IOS-XE  |Yes|
Cisco NXOS (ssh) | Yes|
Cisco NXOS (nxapi)| No|
Arista EOS | No|

For the platforms where SSH auto-detection does not work, the user will need to:

> 1. Manually define a Platform in Nautobot (this will be a one-time task)

Reviewer: ```suggestion
1. Manually define a Platform in Nautobot (this will be a one-time task in order to support any number of devices using this Platform)
```

2. During onboarding, a Port and Platform must explicitly be specified (in addition to the IP and Site)

> There are two ways the NXOS devices can be accessed: via SSH or the `nxapi` feature. 

> > The former will be autodetected; the latter will require the additional steps

Reviewer: This seems a bit redundant to me given the above table.

## IOS and Junos Auto-Created Platforms

The Onboarding Plugin will automatically create Platforms for vendor operating systems where platform auto-detection works.

> The picture below shows the details of auto-created Platforms for *ios* and *junos*.

Reviewer: ```suggestion
The picture below shows the details of auto-created Platforms for `cisco_ios` and `juniper_junos`.
```

> ![ios and junos platforms](docs/images/ios_junos_auto_created_platforms.png)

Reviewer: Relative links to images won't work when the README is published on PyPI. Probably change these to absolute links to `raw.githubusercontent.com`; after this is merged something like `https://raw.githubusercontent.com/nautobot/nautobot-plugin-device-onboarding/main/docs/images/ios_junos_auto_created_platforms.png` should work.

## Creating a Platform

This section will demonstrate how to create a new Platform in the Nautobot UI. Specifically, it will offer examples for
creating Cisco `nxapi` and Arista `eos` Platforms, but the concepts are applicable to any Platform that is manually created.

In the Nautobot dropdown menu, go to Devices--> Platforms--> Add/+.

Define the attributes for the Platform on this screen and click on the 'Create' button. 

> The Slug value will be auto-populated based on the Platform Name, but you can overwrite that auto-populated value.

Reviewer: ```suggestion
The Slug value will be auto-populated based on the Platform Name, but you can overwrite that auto-populated value. For the platform to work correctly with this plugin, in many cases you will need to set a specific Slug value for it to work properly.
```

> By default, the plugin is using the credentials defined in the main `configuration.py` for Napalm (`NAPALM_USERNAME`/`NAPALM_PASSWORD`). It's possible to define specific credentials for each onboarding task.

### Onboarding a Cisco NXOS Device Running the `nxapi` Feature

When onboarding an NXOS device with the `nxapi` feature, there are a few requirements:
* The `Port` must be the same value configured for `nxapi https port` on the Cisco Nexus device
* The `Platform` must be explicitly set to be one with the specific parameters in the [Cisco NXOS Platform](#cisco-nxos-platform) section

> ![onboard nxos](docs/images/onboard_nxos.png)

Reviewer: What's the purpose of the large gray box around these images? It crowds the actual content of the image into a smaller space than it might otherwise be more readable as.

### Onboarding an Arista EOS Device 

When onboarding an Arista EOS device, there are a few requirements:

> * The `Port` must be the same value configured for https on the Arista device

Reviewer: ```suggestion
* The `Port` must be the same value configured for HTTPS on the Arista device
```

* The `Platform` must be explicitly set to be the one with the specific parameters in the [Arista EOS Platform](#arista-eos-platform) section

![onboard eos](docs/images/onboard_eos.png)

## Consult the Status of Onboarding Tasks

The status of the onboarding process for each device is maintained is a dedicated table in Nautobot and can be retrived :
- Via the UI `/plugins/device-onboarding/`
- Via the API `GET /api/plugins​/device-onboarding​/onboarding​/`

## API

> The plugin includes 4 API endpoints to manage the onbarding tasks

Reviewer: ```suggestion
The plugin includes 4 API endpoints to manage the onboarding tasks:
```

DELETE    ​ /api/plugins​/device-onboarding​/onboarding​/{id}​/  Delete a specific onboarding task
```

## Customizing Onboarding Behaviour With Onboarding Extensions

> Onboarding plugin provides methods to customize onboarding behaviour. By creating onboarding extensions, it is possible to onboard switch stacks, HA pairs and perform other customizations. Please see dedicated FAQ for device onboarding: [onboarding_extensions.md](docs/onboarding-extensions/onboarding_extensions.md)

Reviewer: These links should also be changed to absolute links so that they work properly on PyPI.

```suggestion
This plugin provides methods to customize onboarding behaviour. By creating onboarding extensions, it is possible to onboard switch stacks, HA pairs and perform other customizations. Please see the dedicated FAQ for [device onboarding](https://github.com/nautobot/nautobot-plugin-device-onboarding/blob/develop/docs/onboarding-extensions/onboarding_extensions.md).
```
